lib/engine/game/g_1871/game.rb

```diff
@@ -0,0 +1,822 @@
+        SELL_AFTER = :operate
+        SELL_BUY_ORDER = :sell_buy
+        SELL_MOVEMENT = :down_block
+        STARTING_CASH = { 3 => 580, 4 => 460 }.freeze
+        TILE_LAYS = [{ lay: true, upgrade: true }, { lay: :not_if_upgraded, cost: 20, upgrade: false }].freeze
+
+        # This allows us to add in privates to the cert count ourselves, without
+        # the PEIR shares later on
+        CERT_LIMIT_INCLUDES_PRIVATES = false
+
+        # This game only allows you to sell 30 percent of a company at a time.
+        # This is a feature we added to 18xx.games controlled by this variable.
+        TURN_SELL_LIMIT = 30
+
+        # Change the companies text to clearly talk about the Hunslet
+        STATUS_TEXT = {
+          'can_buy_companies' =>
+          ['Can Buy Hunslet', 'Corporation owned by the owner of the Hunslet maybe purchase it from the player'],
+        }.freeze
+
+        # The numbers needed to assign proper private numbers to the PEIR
+        # shares/privates to match the IRL game.
+        NUMBERS = {
+          'So' => 1,
+          'A' => 2,
+          'MS' => 3,
+          'MR' => 4,
+          'S' => 5,
+          'Gt' => 6,
+          'C' => 7,
+        }.freeze
+
+        # In this game most shares start in the market, but if you split a
+        # corporation shares end up in the treasury. We have no IPO.
+        def ipo_name(_entity = nil)
+          'Treasury'
+        end
+
+        # We use the reserved share type to represent the 3 shortline exchange
+        # shares.
+        def ipo_reserved_name(_entity = nil)
+          'Exchange'
+        end
+
+        def setup_preround
+          # Pick a random company to be the mainline and the shortline
+          @corporations = @corporations[0..6].sort_by { self.rand } + @corporations[7..-1]
+          @random_corporation = @corporations[2]
+
+          # Unsort corporations 2-6
+          @corporations = @corporations[0..1] + @corporations[2..6].sort_by { |c| NUMBERS[c.id] } + @corporations[7..-1]
+
+          # Setup the ML / SL names on their privates
+          @companies.each do |company|
+            company.desc = company.desc.gsub('Mainline', mainline.full_name[0..-5])
+            company.desc = company.desc.gsub('Shortline', shortline.full_name[0..-5])
+
+            company.desc = company.desc.gsub('Random', @random_corporation.full_name[0..-5]) if company.id == 'UB'
+          end
+
+          # Setup the ML and SL names, float values and pars
+          mainline.full_name = mainline.full_name[0..-2] + 'ML'
+          mainline.float_percent = 50
+          shortline.full_name = shortline.full_name[0..-2] + 'SL'
+          stock_market.set_par(mainline, stock_market.share_price(1, 1))
+          stock_market.set_par(shortline, stock_market.share_price(2, 1))
+
+          # Mainline and Shortline tokens start on the board
+          @hexes.find { |hex| hex.id == mainline.coordinates }.tile.cities.first.exchange_token(mainline.tokens.first)
```

Comment:
just do one for loop and do this

@hexes.each |hex| do 
  if mainline or shortline...